--- eniac_a/src/atm_phy_schemes/mo_convect_tables.f90	2018-10-15 13:29:39.587939296 +0200
+++ eniac_b/src/atm_phy_schemes/mo_convect_tables.f90	2018-10-15 13:29:20.449567000 +0200
@@ -153,6 +153,16 @@
 
   !------------------------------------------------------------------------------
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+#ifdef __SPLINE_TEST__
+  PUBLIC :: dua, ua, za
+#endif
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 CONTAINS
 
   SUBROUTINE init_convect_tables
