--- eniac_a/src/atm_phy_schemes/mo_convect_tables.f90	2018-05-31 10:02:25.952665384 +0200
+++ eniac_b/src/atm_phy_schemes/mo_convect_tables.f90	2018-05-31 10:02:19.980498000 +0200
@@ -151,6 +151,16 @@
 
   !------------------------------------------------------------------------------
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+#ifdef __SPLINE_TEST__
+  PUBLIC :: dua, ua, za
+#endif
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 CONTAINS
 
   SUBROUTINE init_convect_tables
