--- a/src/atm_phy_echam/mo_vdiff_solver.f90	2018-05-16 13:35:29.309345228 +0200
+++ b/src/atm_phy_echam/mo_vdiff_solver.f90	2018-05-16 13:35:28.242435000 +0200
@@ -52,6 +52,14 @@
 
   CHARACTER(len=*), PARAMETER :: thismodule = 'mo_vdiff_solver'
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+  PUBLIC :: ibtm_mtrx, ibtm_var, ithv, itotte, itrc_start, ixi, ixl, ixv, matrix_idx
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 CONTAINS
   !>
   !!
