--- a/src/configure_model/mo_echam_phy_config.f90	2018-05-16 12:10:43.249187036 +0200
+++ b/src/configure_model/mo_echam_phy_config.f90	2018-05-16 12:10:41.965619000 +0200
@@ -217,6 +217,14 @@
   !!
   TYPE(timedelta) , POINTER :: dt_zero
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+  PUBLIC :: t_echam_phy_config
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 CONTAINS
 
   !----
