--- a/src/atm_phy_echam/mo_surface_diag.f90	2018-05-15 12:36:21.970513000 +0200
+++ b/src/atm_phy_echam/mo_surface_diag.f90	2018-05-15 12:36:42.466031725 +0200
@@ -442,6 +442,7 @@
     REAL(wp), POINTER :: pbtile(:,:)
 #ifdef HAVE_FC_ATTRIBUTE_CONTIGUOUS
     CONTIGUOUS :: pbtile
+#endif
 
   ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
   
@@ -453,7 +454,6 @@
   ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
   
 
-#endif
 
     !$ACC DATA PCOPYIN( pfrc, pqm1, pzf, pzs, pcptgz, pcpt_tile, pbn_tile,     &
     !$ACC               pbhn_tile, pbh_tile, pbm_tile, pri_tile, ptm1, papm1,  &
