diff --git a/run/exp.atm_amip_test b/run/exp.atm_amip_test
index 50a34c9fa..dd9ca71e4 100644
--- a/run/exp.atm_amip_test
+++ b/run/exp.atm_amip_test
@@ -37,12 +37,7 @@ unset OMPI_MCA_coll_fca_priority
 
 # overwrite the default setting with the new path and handle daint (CSCS)
 
-if [ -d /users/icontest ]
-then
-    poolFolder_prefix=/users/icontest
-else
-    poolFolder_prefix=
-fi
+poolFolder_prefix=/project/c15/ICON
 
 icon_data_poolFolder="$poolFolder_prefix/pool/data/ICON/grids/private/mpim/icon_preprocessing/source"
 
