diff --git a/config/mh-linux b/config/mh-linux
index 4db43effa..cc0ac8d65 100644
--- a/config/mh-linux
+++ b/config/mh-linux
@@ -45,7 +45,7 @@ case "$fortran_compiler" in
         compiler_version = $(gfortran --version | awk 'NR==1{split($NF,v,"."); printf("gcc%1d%1d", v[1], v[2])}')
         CFLAGS      = $CFLAGS -DpgiFortran
         FCPP        = -cpp
-        FLANG       = -std=f2008 -fmodule-private -fimplicit-none -fmax-identifier-length=63 -ffree-line-length-132
+        FLANG       = -std=f2008 -fmodule-private -fimplicit-none -fmax-identifier-length=63 -ffree-line-length-none
         FWARN       = -Wall -Wcharacter-truncation -Wconversion -Wunderflow -Wunused-parameter -Wno-surprising
         FDEBUG      = -g -fbacktrace
         FCHECK      = -finit-real=nan -finit-integer=-2147483648 -finit-character=127
@@ -53,7 +53,7 @@ case "$fortran_compiler" in
         FOTPTIM     = -march=native -O3 -ffast-math -D__LOOP_EXCHANGE
 
         DEBUG_FLAGS = -g -fbacktrace -fbounds-check -fstack-protector-all $FCHECK
-        STD_FLAGS   = -march=native -O2 -mpc64 $DEBUG_FLAGS -D__LOOP_EXCHANGE 
+        STD_FLAGS   = -march=native -O2 -mpc64 -D__LOOP_EXCHANGE 
         HIOPT_FLAGS = -march=native -O3 -ffast-math -mpc64 -D__LOOP_EXCHANGE -g
 
         FC          = gfortran
@@ -136,7 +136,7 @@ case "$fortran_compiler" in
         MODDIR      = -mdir 
         WLFLAG      = -Wl,-Wl,
         ;;
-    pgi)
+    pgi*)
         config_compiler  = pgi
         compiler_version = $(pgfortran -V | awk 'NR==2{split($2,v,"."); printf("pgi%1d", v[1])}')
         CFLAGS      = $CFLAGS -DpgiFortran
@@ -238,18 +238,20 @@ case "$ac_sitename" in
                     gcc)
                         config_target=daint_cpu
                         load_modules =modules craype daint-gpu PrgEnv-gnu cray-netcdf cray-hdf5
+                        FFLAGS      = ${FFLAGS} -I\$(SERIALBOX2ROOT)/include
                         CFLAGS      = -DgFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4
                         NETCDFROOT  = ${NETCDF_DIR}
                         HDF5ROOT    = ${HDF5_DIR}
                         LAPACKROOT    = ${CRAY_LIBSCI_PREFIX_DIR}
                         LAPACK_LIB    = -lsci_gnu
-                        OTHER_LIBS  = -L/apps/daint/UES/6.0.UP02/sandbox-ws/jasper/2.0.10/GNU/5.3.0/lib -ljasper -dynamic -L/usr/lib64 -llzma
+                        OTHER_LIBS  = -L/apps/daint/UES/6.0.UP02/sandbox-ws/jasper/2.0.10/GNU/5.3.0/lib -ljasper -dynamic -L/usr/lib64 -llzma -L\$(SERIALBOX2ROOT)/lib -l:libSerialboxFortran.a -lstdc++fs -lstdc++
                         SERIALBOX2ROOT = /project/c14/install/daint/serialbox2/gnu
 			;;
                     intel)
                         config_target=daint_cpu
                         load_modules = PrgEnv-intel cray-netcdf
-                        CFLAGS      = -std=gnu99 -march=native -DpgiFortran -DHAVE_LIBNETCDF
+                        CFLAGS      = -std=gnu99 -march=native -DpgiFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4
+                        FFLAGS      = ${FFLAGS} -I\$(SERIALBOX2ROOT)/include
                         STD_FLAGS   = -g -O2 -qopt-report -fp-model source -fpp -traceback  -D__LOOP_EXCHANGE
                         F77FLAGS    = -g -O2 -qopt-report -fp-model source -fpp -traceback  
                         FFLAGS      = -g -O2 -qopt-report -fp-model source -fpp -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -assume realloc_lhs
@@ -259,34 +261,45 @@ case "$ac_sitename" in
                     pgi)
                         config_target=daint_cpu 
                         load_modules = PrgEnv-pgi cray-netcdf
-                        CFLAGS      = -DpgiFortran -DHAVE_LIBNETCDF
+                        FFLAGS      = ${FFLAGS} -I\$(SERIALBOX2ROOT)/include
+                        CFLAGS      = -DpgiFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4
                         NETCDFROOT  = /opt/cray/pe/netcdf/4.4.1/PGI/15.3
                         SERIALBOX2ROOT = /project/c14/install/daint/serialbox2/pgi
 			;;
+                    pgi_gpu)
+                        config_target=daint_gpu 
+                        load_modules = PrgEnv-pgi cray-netcdf
+                        FFLAGS      = ${FFLAGS} -I\$(SERIALBOX2ROOT)/include -acc=verystrict -ta=tesla,cc60,cuda8.0 -Minfo=accel,inline
+                        CFLAGS      = -DpgiFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4
+                        LDFLAGS     = -ta=tesla,cc60,cuda8.0
+                        NETCDFROOT  = /opt/cray/pe/netcdf/4.4.1/PGI/15.3
+                        SERIALBOX2ROOT = /project/c14/install/daint/serialbox2/pgi
+                        OTHER_LIBS  = \$(SERIALBOX2ROOT)/lib/libSerialboxFortran.a -lstdc++fs -lstdc++
+			                  ;;
                     cray)
                         config_target=daint_cpu
-                        load_modules = modules craype daint-gpu PrgEnv-cray cdt/16.11 cray-netcdf cray-hdf5
-                        FFLAGS      = -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -Df2cFortran -e Z -em -hflex_mp=conservative -hfp1 -hadd_paren -hnoacc -O1,cache0 -r am -Ktrap=divz,ovf,inv
+                        load_modules = modules craype daint-gpu PrgEnv-cray cray-netcdf cray-hdf5
+                        FFLAGS      = -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -Df2cFortran -e Z -em -hflex_mp=conservative -hfp1 -hadd_paren -hnoacc -O1,cache0 -r am -Ktrap=divz,ovf,inv -I\$(SERIALBOX2ROOT)/include
                         CFLAGS      = -g -Df2cFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4 -O3 -DUSE_XML2
                         LDFLAGS     = -Wl,--allow-multiple-definition
                         NETCDFROOT  = ${NETCDF_DIR}
                         HDF5ROOT    = ${HDF5_DIR}
                         LAPACKROOT  = ${CRAY_LIBSCI_PREFIX_DIR}
                         LAPACK_LIB  = -lsci_cray
-                        OTHER_LIBS  = -L/apps/daint/UES/6.0.UP02/sandbox-ws/jasper/2.0.10/CRAY/8.5.5/lib -ljasper -dynamic -L/usr/lib64 -llzma
-                        SERIALBOX2ROOT = /project/c14/install/daint/serialbox2_cce8.5.5
+                        OTHER_LIBS  = -L/apps/daint/UES/6.0.UP02/sandbox-ws/jasper/2.0.10/CRAY/8.5.5/lib -ljasper -dynamic -L/usr/lib64 -llzma -L\$(SERIALBOX2ROOT)/lib -l:libSerialboxFortran.a -lstdc++fs -lstdc++
+                        SERIALBOX2ROOT = /project/c14/install/daint/serialbox2/cray
                         ;;
                     cray_gpu)
                         config_target=daint_gpu
-                        load_modules = modules craype daint-gpu PrgEnv-cray cdt/16.11 cray-netcdf cray-hdf5 craype-accel-nvidia60
-                        FFLAGS      = -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -Df2cFortran -e Z -em -hflex_mp=intolerant -hfp0 -hacc -hnoomp -hadd_paren -hacc_model=deep_copy -O2 -h ipafrom=../../../src/shared/mo_loopindices.f90 -r am -Ktrap=divz,ovf,inv
+                        load_modules = modules craype daint-gpu PrgEnv-cray cray-netcdf cray-hdf5 craype-accel-nvidia60
+                        FFLAGS      = -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -Df2cFortran -e Z -em -hflex_mp=intolerant -hfp0 -hacc -hnoomp -hadd_paren -hacc_model=no_deep_copy -O2 -h ipafrom=../../../src/shared/mo_loopindices.f90 -r am -Ktrap=divz,ovf,inv -I\$(SERIALBOX2ROOT)/include
                         CFLAGS      = -g -Df2cFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF4 -DHAVE_NETCDF2 -O3 -DUSE_XML2
                         NETCDFROOT  = ${NETCDF_DIR}
                         HDF5ROOT    = ${HDF5_DIR}
                         LAPACKROOT  = ${CRAY_LIBSCI_PREFIX_DIR}
                         LAPACK_LIB  = -lsci_cray
-                        OTHER_LIBS  = -L/apps/daint/UES/6.0.UP02/sandbox-ws/jasper/2.0.10/CRAY/8.5.5/lib -ljasper -llzma
-                        SERIALBOX2ROOT = /project/c14/install/daint/serialbox2_cce8.5.5
+                        OTHER_LIBS  = -L/apps/daint/UES/6.0.UP02/sandbox-ws/jasper/2.0.10/CRAY/8.5.5/lib -ljasper -llzma -L\$(SERIALBOX2ROOT)/lib -l:libSerialboxFortran.a -lstdc++fs -lstdc++
+                        SERIALBOX2ROOT = /project/c14/install/daint/serialbox2/cray
 			                  ;;
                     *)
                         echo
@@ -367,7 +380,8 @@ case "$ac_sitename" in
                         MODOPT      = -I
                         MODDIR      = -J
                         config_target = kesch_cpu
-                        FFLAGS        = $FLANG $FWARN -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -D__LOOP_EXCHANGE -DgFortran -cpp -O2
+                        FFLAGS        = $FLANG $FWARN -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -D__LOOP_EXCHANGE -DgFortran -cpp -save-temps -O -g -I\$(SERIALBOX2ROOT)/include -DFORCE_FTG_SERIALIZATION
+                        #FFLAGS        = $FLANG $FWARN -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -D__LOOP_EXCHANGE -DgFortran -cpp -O2 -I\$(SERIALBOX2ROOT)/include
                         CFLAGS        = -D_XOPEN_SOURCE=700 -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -O2 -cpp -std=c99 -DgFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4
                         GRIBAPIROOT    = /project/c14/install/kesch/libgrib_api/v1.20.0.2/gnu/
                         SERIALBOX2ROOT = /project/c14/install/kesch/serialbox2/gnu
@@ -380,7 +394,7 @@ case "$ac_sitename" in
                                 HDF5ROOT      = ${EBROOTHDF5}
                                 LAPACKROOT    = ${CRAY_LIBSCI_PREFIX_DIR}
                                 LAPACK_LIB    = -lsci_cray
-                                OTHER_LIBS    = -lpthread
+                                OTHER_LIBS    = -lpthread -L\$(SERIALBOX2ROOT)/lib -l:libSerialboxFortran.a -lstdc++fs -lstdc++
                                 ;;
                             *)
                                 (>&6 echo "")
@@ -398,18 +412,35 @@ case "$ac_sitename" in
                         DEFCOPT     = -D
                         MODOPT      = -I
                         MODDIR      = -module 
-                        config_target = kesch_cpu
-                        FFLAGS        = -Mpreprocess -Mrecursive -Mallocatable=03 -D__LOOP_EXCHANGE $INCLUDES
+                        FFLAGS        = -Mpreprocess -Mrecursive -Mallocatable=03 -D__LOOP_EXCHANGE $INCLUDES -I\$(SERIALBOX2ROOT)/include
                         CFLAGS        = -c99 -DpgiFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4
                         GRIBAPIROOT    = /project/c14/install/kesch/libgrib_api/v1.20.0.2/pgi/
                         SERIALBOX2ROOT = /project/c14/install/kesch/serialbox2/pgi
+                        OTHER_LIBS    = \$(SERIALBOX2ROOT)/lib/libSerialboxFortran.a -lstdc++fs -lstdc++
                         config_tuning = $(tmp=${fortran_compiler#'pgi'};tmp=${tmp#_};tmp=${tmp:-default};echo $tmp)
+                        # Generic CPU/GPU setup
+                        case "${config_tuning}" in
+                            gpu*)
+                                config_target = kesch_gpu
+                                FFLAGS        = ${FFLAGS} -acc=verystrict -ta=tesla,cc35,cuda8.0 -Minfo=accel,inline
+                                LDFLAGS       = -ta=tesla,cc35,cuda8.0
+                                ;;
+                            *)
+                                config_target = kesch_cpu
+                                FFLAGS        = ${FFLAGS}
+                                ;;
+                        esac
+                        # Setup of specific version
                         case "${config_tuning}" in
                             default)
-                                load_modules  = git/2.14.1 craype-haswell PrgEnv-gnu netcdf netcdf-fortran hdf5 cray-libsci
-                                NETCDFROOT    = ${EBROOTNETCDF}
-                                NETCDFFROOT   = ${EBROOTNETCDFMINFORTRAN}
-                                HDF5ROOT      = ${EBROOTHDF5}
+                                load_modules  = git/2.14.1 PrgEnv-pgi pgi/17.10 gcc/5.4.0-2.26 mvapich2gdr_gnu/2.3a_cuda_8.0_pgi17.10 netcdf/4.4.1.1/2017
+                                NETCDFROOT    = ${NETCDF_ROOT}
+                                HDF5ROOT      = ${NETCDF_ROOT}
+                                ;;
+                            gpu)
+                                load_modules  = git/2.14.1 PrgEnv-pgi pgi/17.10 gcc/5.4.0-2.26 mvapich2gdr_gnu/2.3a_cuda_8.0_pgi17.10 netcdf/4.4.1.1/2017
+                                NETCDFROOT    = ${NETCDF_ROOT}
+                                HDF5ROOT      = ${NETCDF_ROOT}
                                 ;;
                             *)
                                 (>&6 echo "")
@@ -429,18 +460,19 @@ case "$ac_sitename" in
                         MODOPT      = -I
                         MODDIR      = -J
                         # FDEBUG        = -O0 -g -Rb -Rc -Rd -Rp #-Rs leads to ICE
-                        FFLAGS        = -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -Df2cFortran -e Z -em -hadd_paren -r am -Ktrap=divz,ovf,inv
+                        FFLAGS        = -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -D__LOOP_EXCHANGE -D__MIXED_PRECISION -D__MIXED_PRECISION_2 -Df2cFortran -e Z -em -hadd_paren -r am -Ktrap=divz,ovf,inv -I\$(SERIALBOX2ROOT)/include
                         # CDEBUG       = -O0 -g
                         CFLAGS        = -DHAVE_SLOW_PASSIVE_TARGET_ONESIDED -Df2cFortran -DHAVE_LIBNETCDF -DHAVE_NETCDF2 -DHAVE_NETCDF4 -O3 -DUSE_XML2
                         LDFLAGS       = -Wl,--allow-multiple-definition
                         GRIBAPIROOT    = /project/c14/install/kesch/libgrib_api/v1.20.0.2/cray/
                         SERIALBOX2ROOT = /project/c14/install/kesch/serialbox2/cray
+                        OTHER_LIBS    = -L\$(SERIALBOX2ROOT)/lib -l:libSerialboxFortran.a -lstdc++fs -lstdc++
                         config_tuning = $(tmp=${fortran_compiler#'cray'};tmp=${tmp#_};tmp=${tmp:-default};echo $tmp)
                         # Generic CPU/GPU setup
                         case "${config_tuning}" in
                             gpu*)
                                 config_target = kesch_gpu
-                                FFLAGS        = ${FFLAGS} -hnoomp -hflex_mp=intolerant -hfp0 -hacc_model=deep_copy -Ocache0,ipa1
+                                FFLAGS        = ${FFLAGS} -hnoomp -hflex_mp=intolerant -hfp1 -hacc_model=no_deep_copy -Ocache0,ipa1
                                 ;;
                             *)
                                 config_target = kesch_cpu
