--- eniac_a/src/atm_phy_schemes/mo_convect_tables.f90	2018-10-15 13:20:20.000000000 +0200
+++ eniac_b/src/atm_phy_schemes/mo_convect_tables.f90	2018-10-15 13:20:25.391026000 +0200
@@ -156,7 +156,9 @@
 
   ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
   
+#ifdef __SPLINE_TEST__
   PUBLIC :: dua, ua, za
+#endif
   
   ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
   
