--- eniac_a/src/atm_phy_echam/mo_vdiff_downward_sweep.f90	2018-10-31 09:17:32.705569971 +0100
+++ eniac_b/src/atm_phy_echam/mo_vdiff_downward_sweep.f90	2018-10-31 09:24:16.111549282 +0100
@@ -526,8 +526,10 @@
   &  mo_vdiff_solver__ibtm_var => ibtm_var, mo_vdiff_solver__ithv => ithv, mo_vdiff_solver__iqv => iqv, mo_vdiff_solver__ibtm_mtrx &
   &  => ibtm_mtrx, mo_vdiff_solver__ih => ih, mo_vdiff_solver__itrc_start => itrc_start, mo_vdiff_solver__iv => iv, &
   &  mo_vdiff_solver__iu => iu
-  USE mo_convect_tables, ONLY: mo_convect_tables__za => za, mo_convect_tables__ua => ua, mo_convect_tables__tlucu => tlucu, &
-  &  mo_convect_tables__lookupoverflow => lookupoverflow, mo_convect_tables__dua => dua
+  USE mo_convect_tables, ONLY: mo_convect_tables__tlucu => tlucu, mo_convect_tables__lookupoverflow => lookupoverflow
+#ifdef __SPLINE_TEST__
+  USE mo_convect_tables, ONLY: mo_convect_tables__za => za, mo_convect_tables__ua => ua, mo_convect_tables__dua => dua
+#endif
   USE mo_model_domain, ONLY: mo_model_domain__p_patch => p_patch, t_patch
   USE mo_echam_vdiff_params, ONLY: mo_echam_vdiff_params__itop => itop
   
@@ -950,11 +952,13 @@
     
     
     ! GLOBALS
+#ifdef __SPLINE_TEST__
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_convect_tables__dua)
     !$ACC UPDATE HOST( mo_convect_tables__dua )
 #endif
     
     CALL ftg_write("mo_convect_tables__dua", mo_convect_tables__dua, LBOUND(mo_convect_tables__dua), UBOUND(mo_convect_tables__dua))
+#endif
     IF (ALLOCATED(mo_vdiff_solver__ibtm_mtrx)) THEN
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_vdiff_solver__ibtm_mtrx)
       !$ACC UPDATE HOST( mo_vdiff_solver__ibtm_mtrx )
@@ -1001,6 +1005,7 @@
     
     CALL ftg_write("mo_convect_tables__tlucu", mo_convect_tables__tlucu, LBOUND(mo_convect_tables__tlucu), UBOUND( &
     &  mo_convect_tables__tlucu))
+#ifdef __SPLINE_TEST__
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_convect_tables__ua)
     !$ACC UPDATE HOST( mo_convect_tables__ua )
 #endif
@@ -1011,6 +1016,7 @@
 #endif
     
     CALL ftg_write("mo_convect_tables__za", mo_convect_tables__za, LBOUND(mo_convect_tables__za), UBOUND(mo_convect_tables__za))
+#endif
     CALL ftg_register_only("mo_echam_vdf_config__echam_vdf_config", "TYPE(t_echam_vdf_config)", LBOUND( &
     &  mo_echam_vdf_config__echam_vdf_config), UBOUND(mo_echam_vdf_config__echam_vdf_config))
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_echam_vdf_config__echam_vdf_config_c_e)
@@ -1198,8 +1204,10 @@
   &  mo_vdiff_solver__ibtm_var => ibtm_var, mo_vdiff_solver__ithv => ithv, mo_vdiff_solver__iqv => iqv, mo_vdiff_solver__ibtm_mtrx &
   &  => ibtm_mtrx, mo_vdiff_solver__ih => ih, mo_vdiff_solver__itrc_start => itrc_start, mo_vdiff_solver__iv => iv, &
   &  mo_vdiff_solver__iu => iu
-  USE mo_convect_tables, ONLY: mo_convect_tables__za => za, mo_convect_tables__ua => ua, mo_convect_tables__tlucu => tlucu, &
-  &  mo_convect_tables__lookupoverflow => lookupoverflow, mo_convect_tables__dua => dua
+  USE mo_convect_tables, ONLY: mo_convect_tables__tlucu => tlucu, mo_convect_tables__lookupoverflow => lookupoverflow
+#ifdef __SPLINE_TEST__
+  USE mo_convect_tables, ONLY: mo_convect_tables__za => za, mo_convect_tables__ua => ua, mo_convect_tables__dua => dua
+#endif
   USE mo_model_domain, ONLY: mo_model_domain__p_patch => p_patch, t_patch
   USE mo_echam_vdiff_params, ONLY: mo_echam_vdiff_params__itop => itop
   
@@ -1408,11 +1416,13 @@
     
     
     ! GLOBALS
+#ifdef __SPLINE_TEST__
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_convect_tables__dua)
     !$ACC UPDATE HOST( mo_convect_tables__dua )
 #endif
     
     CALL ftg_write("mo_convect_tables__dua", mo_convect_tables__dua, LBOUND(mo_convect_tables__dua), UBOUND(mo_convect_tables__dua))
+#endif
     IF (ALLOCATED(mo_vdiff_solver__ibtm_mtrx)) THEN
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_vdiff_solver__ibtm_mtrx)
       !$ACC UPDATE HOST( mo_vdiff_solver__ibtm_mtrx )
@@ -1459,6 +1469,7 @@
     
     CALL ftg_write("mo_convect_tables__tlucu", mo_convect_tables__tlucu, LBOUND(mo_convect_tables__tlucu), UBOUND( &
     &  mo_convect_tables__tlucu))
+#ifdef __SPLINE_TEST__
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_convect_tables__ua)
     !$ACC UPDATE HOST( mo_convect_tables__ua )
 #endif
@@ -1469,6 +1480,7 @@
 #endif
     
     CALL ftg_write("mo_convect_tables__za", mo_convect_tables__za, LBOUND(mo_convect_tables__za), UBOUND(mo_convect_tables__za))
+#endif
     CALL ftg_register_only("mo_echam_vdf_config__echam_vdf_config", "TYPE(t_echam_vdf_config)", LBOUND( &
     &  mo_echam_vdf_config__echam_vdf_config), UBOUND(mo_echam_vdf_config__echam_vdf_config))
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_mo_echam_vdf_config__echam_vdf_config_c_e)
