--- eniac_a/src/parallel_infrastructure/mo_sync.f90	2018-09-25 11:29:16.919191065 +0200
+++ eniac_b/src/parallel_infrastructure/mo_sync.f90	2018-09-25 11:29:09.696147000 +0200
@@ -159,6 +159,14 @@
 #endif
 #endif
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+  PUBLIC :: do_sync_checks, log_unit
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 CONTAINS
 
 
