--- eniac_a/src/atm_phy_schemes/gscp_graupel.f90	2018-10-29 06:58:20.000000000 +0100
+++ eniac_b/src/atm_phy_schemes/gscp_graupel.f90	2018-10-29 07:02:01.108420025 +0100
@@ -607,18 +607,6 @@
 
   LOGICAL :: ldum
 
-  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
-  
-  ftg_graupel_round = ftg_graupel_round + 1
-  CALL ftg_graupel_capture_input(nvec, ke, ivstart, ivend, kstart, idbg, zdt, dz, t, p, rho, qv, qc, qi, qr, qs, qg, qnc, qi0, &
-  &  qc0, prr_gsp, prs_gsp, prg_gsp, tinc_lh, pstoph, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, &
-  &  ddt_tend_qc, ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, ddt_diag_idep, &
-  &  ddt_diag_sdep, ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, ddt_diag_imelt, &
-  &  ddt_diag_smelt, ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
-  
-  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
-  
-
 
 !------------ End of header ---------------------------------------------------
 
@@ -640,6 +628,18 @@
   fxna_cooper(ztx) = 5.0E+0_wp * EXP(0.304_wp * (t0 - ztx))   ! FR: Cooper (1986) used by Greg Thompson(2008)
 
 
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+  ftg_graupel_round = ftg_graupel_round + 1
+  CALL ftg_graupel_capture_input(nvec, ke, ivstart, ivend, kstart, idbg, zdt, dz, t, p, rho, qv, qc, qi, qr, qs, qg, qnc, qi0, &
+  &  qc0, prr_gsp, prs_gsp, prg_gsp, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, &
+  &  ddt_tend_qc, ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, ddt_diag_idep, &
+  &  ddt_diag_sdep, ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, ddt_diag_imelt, &
+  &  ddt_diag_smelt, ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 ! Define reciprocal of heat capacity of dry air (at constant pressure vs at constant volume)
 
 #ifdef __COSMO__
@@ -1772,7 +1772,7 @@
 
   ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
   
-  CALL ftg_graupel_capture_output(t, qv, qc, qi, qr, qs, qg, qnc, prr_gsp, prs_gsp, prg_gsp, tinc_lh, qrsflux, ddt_tend_t, &
+  CALL ftg_graupel_capture_output(t, qv, qc, qi, qr, qs, qg, qnc, prr_gsp, prs_gsp, prg_gsp, qrsflux, ddt_tend_t, &
   &  ddt_tend_qv, ddt_tend_qc, ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, &
   &  ddt_diag_idep, ddt_diag_sdep, ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, &
   &  ddt_diag_imelt, ddt_diag_smelt, ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
@@ -1824,7 +1824,7 @@
 END FUNCTION ftg_graupel_capture_input_active
 
 SUBROUTINE ftg_graupel_capture_input(nvec, ke, ivstart, ivend, kstart, idbg, zdt, dz, t, p, rho, qv, qc, qi, qr, qs, qg, qnc, qi0, &
-&  qc0, prr_gsp, prs_gsp, prg_gsp, tinc_lh, pstoph, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, ddt_tend_qc, &
+&  qc0, prr_gsp, prs_gsp, prg_gsp, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, ddt_tend_qc, &
 &  ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, ddt_diag_idep, ddt_diag_sdep, &
 &  ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, ddt_diag_imelt, ddt_diag_smelt, &
 &  ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
@@ -1865,8 +1865,6 @@
   REAL(KIND=wp), DIMENSION(:), INTENT(in) :: prr_gsp
   REAL(KIND=wp), DIMENSION(:), INTENT(in) :: prs_gsp
   REAL(KIND=wp), DIMENSION(:), INTENT(in) :: prg_gsp
-  REAL(KIND=wp), DIMENSION(:,:), INTENT(in) :: tinc_lh
-  REAL(KIND=wp), DIMENSION(:,:), INTENT(in) :: pstoph
   REAL(KIND=wp), DIMENSION(:,:), INTENT(in) :: qrsflux
   LOGICAL, INTENT(in), OPTIONAL :: l_cv
   LOGICAL, INTENT(in), OPTIONAL :: ldiag_ttend
@@ -1973,14 +1971,6 @@
     !$ACC UPDATE HOST(prg_gsp)
 #endif
     CALL ftg_write("prg_gsp", prg_gsp, LBOUND(prg_gsp), UBOUND(prg_gsp))
-#if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_tinc_lh)
-    !$ACC UPDATE HOST(tinc_lh)
-#endif
-    CALL ftg_write("tinc_lh", tinc_lh, LBOUND(tinc_lh), UBOUND(tinc_lh))
-#if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_pstoph)
-    !$ACC UPDATE HOST(pstoph)
-#endif
-    CALL ftg_write("pstoph", pstoph, LBOUND(pstoph), UBOUND(pstoph))
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_qrsflux)
     !$ACC UPDATE HOST(qrsflux)
 #endif
@@ -2209,7 +2199,7 @@
 
 END FUNCTION ftg_graupel_capture_output_active
 
-SUBROUTINE ftg_graupel_capture_output(t, qv, qc, qi, qr, qs, qg, qnc, prr_gsp, prs_gsp, prg_gsp, tinc_lh, qrsflux, ddt_tend_t, &
+SUBROUTINE ftg_graupel_capture_output(t, qv, qc, qi, qr, qs, qg, qnc, prr_gsp, prs_gsp, prg_gsp, qrsflux, ddt_tend_t, &
 &  ddt_tend_qv, ddt_tend_qc, ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, &
 &  ddt_diag_idep, ddt_diag_sdep, ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, &
 &  ddt_diag_imelt, ddt_diag_smelt, ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
@@ -2238,7 +2228,6 @@
   REAL(KIND=wp), DIMENSION(:), INTENT(in) :: prr_gsp
   REAL(KIND=wp), DIMENSION(:), INTENT(in) :: prs_gsp
   REAL(KIND=wp), DIMENSION(:), INTENT(in) :: prg_gsp
-  REAL(KIND=wp), DIMENSION(:,:), INTENT(in) :: tinc_lh
   REAL(KIND=wp), DIMENSION(:,:), INTENT(in) :: qrsflux
   REAL(KIND=wp), DIMENSION(:,:), INTENT(in), OPTIONAL :: ddt_tend_t
   REAL(KIND=wp), DIMENSION(:,:), INTENT(in), OPTIONAL :: ddt_tend_qv
@@ -2325,10 +2314,6 @@
     !$ACC UPDATE HOST(prg_gsp)
 #endif
     CALL ftg_write("prg_gsp", prg_gsp, LBOUND(prg_gsp), UBOUND(prg_gsp))
-#if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_tinc_lh)
-    !$ACC UPDATE HOST(tinc_lh)
-#endif
-    CALL ftg_write("tinc_lh", tinc_lh, LBOUND(tinc_lh), UBOUND(tinc_lh))
 #if defined(FTG_ACC_COPYOUT) && !defined(FTG_ACC_NOCOPYOUT_qrsflux)
     !$ACC UPDATE HOST(qrsflux)
 #endif
