--- src/atm_phy_schemes/mo_convect_tables.f90.orig	2018-05-09 14:37:26.067698000 +0200
+++ src/atm_phy_schemes/mo_convect_tables.f90	2018-05-09 14:37:41.210882185 +0200
@@ -154,7 +154,9 @@
 
   ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
   
+#ifdef __SPLINE_TEST__
   PUBLIC :: dua, ua, za
+#endif
   
   ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
   
