--- eniac_a/src/atm_phy_echam/mo_vdiff_solver.f90	2018-10-30 11:43:53.263744165 +0100
+++ eniac_b/src/atm_phy_echam/mo_vdiff_solver.f90	2018-10-30 11:43:43.264175000 +0100
@@ -53,6 +53,14 @@
 
   CHARACTER(len=*), PARAMETER :: thismodule = 'mo_vdiff_solver'
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+  
+  PUBLIC :: ithv, itotte, itrc_start, ixi, ixl, ixv
+  
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+  
+
 CONTAINS
   !>
   !!
