--- eniac_a/src/atm_phy_echam/mo_vdiff_solver.f90	2018-05-29 17:29:12.000000000 +0200
+++ eniac_b/src/atm_phy_echam/mo_vdiff_solver.f90	2018-05-30 08:47:58.899170598 +0200
@@ -55,7 +55,7 @@
 
   ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
   
-  PUBLIC :: ibtm_var, matrix_idx
+  PUBLIC :: ibtm_var, ithv, itotte, itrc_start, ixi, ixl, ixv, matrix_idx
   
   ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
   
