--- eniac_a/src/atm_phy_echam/mo_vdiff_upward_sweep.f90	2018-10-15 11:49:12.957520000 +0200
+++ eniac_b/src/atm_phy_echam/mo_vdiff_upward_sweep.f90	2018-10-15 12:25:59.735885494 +0200
@@ -202,7 +202,7 @@
   USE mo_impl_constants, ONLY: MAX_CHAR_LENGTH
   USE mo_mpi,            ONLY: get_my_mpi_all_id
   USE m_ser_ftg,         ONLY: ftg_set_serializer, ftg_add_serializer_metainfo, ignore_bullshit_max_dim_size, &
-  &  ignore_bullshit_allow_negative_indices
+  &  ignore_bullshit_allow_negative_indices, ignore_bullshit
   
   CHARACTER(*), INTENT(IN) :: stage, dir
   
@@ -218,6 +218,7 @@
     !    WRITE (0,*) 'FTG INIT *** ERROR: ', TRIM(mkdirerr)
     !  END IF
     
+    ignore_bullshit = .FALSE.
     ignore_bullshit_max_dim_size = 100000
     ignore_bullshit_allow_negative_indices = .TRUE.
     
