--- eniac_a/src/tests/ftg_graupel_test.f90	2018-10-29 07:07:57.000000000 +0100
+++ eniac_b/src/tests/ftg_graupel_test.f90	2018-10-29 07:09:54.593556081 +0100
@@ -69,8 +69,6 @@
     REAL(KIND=wp), DIMENSION(:), ALLOCATABLE :: prr_gsp
     REAL(KIND=wp), DIMENSION(:), ALLOCATABLE :: prs_gsp
     REAL(KIND=wp), DIMENSION(:), ALLOCATABLE :: prg_gsp
-    REAL(KIND=wp), DIMENSION(:,:), ALLOCATABLE :: tinc_lh
-    REAL(KIND=wp), DIMENSION(:,:), ALLOCATABLE :: pstoph
     REAL(KIND=wp), DIMENSION(:,:), ALLOCATABLE :: qrsflux
     LOGICAL :: l_cv
     LOGICAL :: ldiag_ttend
@@ -109,7 +107,7 @@
     
     CALL ftg_graupel_init_for_replay('input')
     CALL ftg_graupel_replay_input(nvec, ke, ivstart, ivend, kstart, idbg, zdt, dz, t, p, rho, qv, qc, qi, qr, qs, qg, qnc, qi0, &
-    &  qc0, prr_gsp, prs_gsp, prg_gsp, tinc_lh, pstoph, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, &
+    &  qc0, prr_gsp, prs_gsp, prg_gsp, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, &
     &  ddt_tend_qc, ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, ddt_diag_idep, &
     &  ddt_diag_sdep, ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, ddt_diag_imelt, &
     &  ddt_diag_smelt, ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
@@ -117,7 +115,7 @@
     
     CALL start_loc_timing("graupel", 1)
     CALL graupel(nvec, ke, ivstart, ivend, kstart, idbg, zdt, dz, t, p, rho, qv, qc, qi, qr, qs, qg, qnc, qi0, qc0, prr_gsp, &
-    &  prs_gsp, prg_gsp, tinc_lh, pstoph, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, ddt_tend_qc, &
+    &  prs_gsp, prg_gsp, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, ddt_tend_qc, &
     &  ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, ddt_diag_idep, ddt_diag_sdep, &
     &  ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, ddt_diag_imelt, ddt_diag_smelt, &
     &  ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
@@ -146,7 +144,7 @@
   END SUBROUTINE ftg_graupel_init_for_replay
   
   SUBROUTINE ftg_graupel_replay_input(nvec, ke, ivstart, ivend, kstart, idbg, zdt, dz, t, p, rho, qv, qc, qi, qr, qs, qg, qnc, &
-  &  qi0, qc0, prr_gsp, prs_gsp, prg_gsp, tinc_lh, pstoph, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, &
+  &  qi0, qc0, prr_gsp, prs_gsp, prg_gsp, qrsflux, l_cv, ldiag_ttend, ldiag_qtend, ddt_tend_t, ddt_tend_qv, &
   &  ddt_tend_qc, ddt_tend_qi, ddt_tend_qr, ddt_tend_qs, ddt_diag_au, ddt_diag_ac, ddt_diag_ev, ddt_diag_nuc, ddt_diag_idep, &
   &  ddt_diag_sdep, ddt_diag_agg, ddt_diag_rim, ddt_diag_rcri, ddt_diag_icri, ddt_diag_dau, ddt_diag_iau, ddt_diag_imelt, &
   &  ddt_diag_smelt, ddt_diag_cfrz, ddt_diag_rfrz, ddt_diag_shed, ddt_tend_qg)
@@ -174,8 +172,6 @@
     REAL(KIND=wp), DIMENSION(:), ALLOCATABLE, INTENT(inout) :: prr_gsp
     REAL(KIND=wp), DIMENSION(:), ALLOCATABLE, INTENT(inout) :: prs_gsp
     REAL(KIND=wp), DIMENSION(:), ALLOCATABLE, INTENT(inout) :: prg_gsp
-    REAL(KIND=wp), DIMENSION(:,:), ALLOCATABLE, INTENT(inout) :: tinc_lh
-    REAL(KIND=wp), DIMENSION(:,:), ALLOCATABLE, INTENT(inout) :: pstoph
     REAL(KIND=wp), DIMENSION(:,:), ALLOCATABLE, INTENT(inout) :: qrsflux
     LOGICAL, INTENT(inout), OPTIONAL :: l_cv
     LOGICAL, INTENT(inout), OPTIONAL :: ldiag_ttend
@@ -280,14 +276,6 @@
 #if defined(FTG_ACC_COPYIN) && !defined(FTG_ACC_NOCOPYIN_prg_gsp)
     !$ACC ENTER DATA COPYIN( prg_gsp )
 #endif
-    CALL ftg_allocate_and_read_allocatable("tinc_lh", tinc_lh, ftg_rperturb)
-#if defined(FTG_ACC_COPYIN) && !defined(FTG_ACC_NOCOPYIN_tinc_lh)
-    !$ACC ENTER DATA COPYIN( tinc_lh )
-#endif
-    CALL ftg_allocate_and_read_allocatable("pstoph", pstoph, ftg_rperturb)
-#if defined(FTG_ACC_COPYIN) && !defined(FTG_ACC_NOCOPYIN_pstoph)
-    !$ACC ENTER DATA COPYIN( pstoph )
-#endif
     CALL ftg_allocate_and_read_allocatable("qrsflux", qrsflux, ftg_rperturb)
 #if defined(FTG_ACC_COPYIN) && !defined(FTG_ACC_NOCOPYIN_qrsflux)
     !$ACC ENTER DATA COPYIN( qrsflux )
