--- eniac_a/src/base/mo_jsb_control.f90	2018-09-12 14:27:13.791740000 +0200
+++ eniac_b/src/base/mo_jsb_control.f90	2018-09-12 14:27:02.209331000 +0200
@@ -28,6 +28,11 @@
   PUBLIC :: t_jsb_model_nml, jsb_models_nml, get_no_of_models
   PUBLIC :: timer_jsbach, timer_aggregate, timer_integrate, timer_process_task, timer_process_msg
   PUBLIC :: max_no_of_models
+  !================= START FORCE FORTRAN TEST GENERATOR (FTG) =======================!
+
+  PUBLIC :: is_standalone, debug_level, timer_level
+
+  !================= STOP FORCE FORTRAN TEST GENERATOR (FTG) =========================!
 
   LOGICAL :: is_standalone
   LOGICAL :: restart_jsbach
