--- eniac_a/src/tests/ftg_update_surface_test.f90	2018-05-29 15:07:48.000000000 +0200
+++ eniac_b/src/tests/ftg_update_surface_test.f90	2018-05-29 15:07:51.245767000 +0200
@@ -171,6 +171,26 @@
     &  albvisdir_ice, albvisdif_ice, albnirdir_ice, albnirdif_ice)
     CALL ftg_destroy_serializer()
     
+    !$ACC ENTER DATA COPYIN( mo_echam_phy_config__echam_phy_config )
+    !$ACC ENTER DATA COPYIN( pfrc, pcfh_tile, pcfm_tile, pfac_sfc, pocu, pocv, &
+    !$ACC                    aa, aa_btm, bb, bb_btm, pcpt_tile, pqsat_tile,    &
+    !$ACC                    ptsfc_tile, pu_stress_gbm, pv_stress_gbm,         &
+    !$ACC                    plhflx_gbm, pshflx_gbm, pevap_gbm,                &
+    !$ACC                    pu_stress_tile, pv_stress_tile, plhflx_tile,      &
+    !$ACC                    pshflx_tile, pevap_tile, pco2nat, lsm, alake, pu, &
+    !$ACC                    pv, ptemp, pq, pco2, prsfl, prsfc, pssfl, pssfc,  &
+    !$ACC                    rlds, rlus, rsds, rsus, rvds_dir, rpds_dir,       &
+    !$ACC                    rnds_dir, rvds_dif, rpds_dif, rnds_dif, ps,       &
+    !$ACC                    pcosmu0, pch_tile, pcsat, pcair, q_snocpymlt,     &
+    !$ACC                    z0m_tile, z0h_lnd, albvisdir, albnirdir,          &
+    !$ACC                    albvisdif, albnirdif, albvisdir_tile,             &
+    !$ACC                    albnirdir_tile, albvisdif_tile, albnirdif_tile,   &
+    !$ACC                    albedo, albedo_tile, pco2_flux_tile, ptsfc,       &
+    !$ACC                    ptsfc_rad, rsns_tile, rlns_tile, lake_ice_frc,    &
+    !$ACC                    Tsurf, T1, T2, hi, hs, Qtop, Qbot, conc,          &
+    !$ACC                    albvisdir_ice, albvisdif_ice, albnirdir_ice,      &
+    !$ACC                    albnirdif_ice )
+
     CALL update_surface(jg, kproma, kbdim, kice, klev, ksfc_type, idx_wtr, idx_ice, idx_lnd, pdtime, pfrc, pcfh_tile, pcfm_tile, &
     &  pfac_sfc, pocu, pocv, aa, aa_btm, bb, bb_btm, pcpt_tile, pqsat_tile, ptsfc_tile, pu_stress_gbm, pv_stress_gbm, plhflx_gbm, &
     &  pshflx_gbm, pevap_gbm, pu_stress_tile, pv_stress_tile, plhflx_tile, pshflx_tile, pevap_tile, pco2nat, nblock, lsm, alake, &
@@ -180,6 +200,26 @@
     &  ptsfc_rad, rsns_tile, rlns_tile, lake_ice_frc, Tsurf, T1, T2, hi, hs, Qtop, Qbot, conc, albvisdir_ice, albvisdif_ice, &
     &  albnirdir_ice, albnirdif_ice)
     
+    !$ACC EXIT DATA DELETE( pfrc, pcfh_tile, pcfm_tile, pfac_sfc, pocu, pocv,  &
+    !$ACC                   aa, aa_btm, bb, bb_btm, pcpt_tile, pqsat_tile,     &
+    !$ACC                   ptsfc_tile, pu_stress_gbm, pv_stress_gbm,          &
+    !$ACC                   plhflx_gbm, pshflx_gbm, pevap_gbm,                 &
+    !$ACC                   pu_stress_tile, pv_stress_tile, plhflx_tile,       &
+    !$ACC                   pshflx_tile, pevap_tile, pco2nat, lsm, alake, pu,  &
+    !$ACC                   pv, ptemp, pq, pco2, prsfl, prsfc, pssfl, pssfc,   &
+    !$ACC                   rlds, rlus, rsds, rsus, rvds_dir, rpds_dir,        &
+    !$ACC                   rnds_dir, rvds_dif, rpds_dif, rnds_dif, ps,        &
+    !$ACC                   pcosmu0, pch_tile, pcsat, pcair, q_snocpymlt,      &
+    !$ACC                   z0m_tile, z0h_lnd, albvisdir, albnirdir,           &
+    !$ACC                   albvisdif, albnirdif, albvisdir_tile,              &
+    !$ACC                   albnirdir_tile, albvisdif_tile, albnirdif_tile,    &
+    !$ACC                   albedo, albedo_tile, pco2_flux_tile, ptsfc,        &
+    !$ACC                   ptsfc_rad, rsns_tile, rlns_tile, lake_ice_frc,     &
+    !$ACC                   Tsurf, T1, T2, hi, hs, Qtop, Qbot, conc,           &
+    !$ACC                   albvisdir_ice, albvisdif_ice, albnirdir_ice,       &
+    !$ACC                   albnirdif_ice )
+    !$ACC EXIT DATA DELETE( mo_echam_phy_config__echam_phy_config )
+    
   END SUBROUTINE ftg_test_update_surface
   
   
@@ -207,6 +247,15 @@
   &  albnirdir, albvisdif, albnirdif, albvisdir_tile, albnirdir_tile, albvisdif_tile, albnirdif_tile, albedo, albedo_tile, &
   &  pco2_flux_tile, ptsfc, ptsfc_rad, rsns_tile, rlns_tile, lake_ice_frc, Tsurf, T1, T2, hi, hs, Qtop, Qbot, conc, albvisdir_ice, &
   &  albvisdif_ice, albnirdir_ice, albnirdif_ice)
+  !=================== START MANUALLY ADDED FOR FTG =====================!
+  USE mo_time_config,    ONLY: set_tc_startdate, set_tc_stopdate, &
+                               set_tc_dt_model, set_tc_current_date, &
+                               set_tc_exp_startdate, set_tc_exp_stopdate
+  USE mo_jsb_base,       ONLY: jsbach_setup_models, jsbach_setup_tiles
+  USE mo_jsb_model_init, ONLY: jsbach_setup_grid, jsbach_init
+  USE mo_echam_convect_tables, ONLY: init_convect_tables
+  USE mo_zaxis_type, ONLY: zaxisTypeList, t_zaxisTypeList
+  !===================== END MANUALLY ADDED FOR FTG =====================!
     
     INTEGER, INTENT(inout) :: jg
     INTEGER, INTENT(inout) :: kproma
@@ -719,6 +768,31 @@
       &  send_startidx)
     END DO
     
+    !===================== START MANUALLY ADDED FOR FTG ===================!
+    mo_run_config__ltimer = .FALSE.
+    CALL init_convect_tables()
+    mo_master_control__master_namelist_filename = "icon_master.namelist"
+    mo_model_domain__p_patch%grid_filename = 'icon_grid_0005_R02B04_G.nc'
+    CALL set_tc_startdate('1979-01-01T00:00:00Z')
+    CALL set_tc_stopdate('1979-01-01T00:16:00Z')
+    CALL set_tc_exp_startdate('1979-01-01T00:00:00Z')
+    CALL set_tc_exp_stopdate('1979-01-01T00:16:00Z')
+    CALL set_tc_current_date('1979-01-01T00:00:00Z')
+    CALL set_tc_dt_model('PT4M')
+
+    ! Do basic initialization of JSBACH
+    zaxisTypeList = t_zaxisTypeList()
+    CALL jsbach_setup_models(mo_master_control__master_namelist_filename)
+    mo_jsb_control__l_timer = .FALSE.
+
+    ! Now continue initialization of JSBACH for the different grids
+    IF (mo_echam_phy_config__echam_phy_config(jg)%ljsb) THEN 
+      CALL jsbach_setup_grid(jg, mo_model_domain__p_patch(jg))
+      CALL jsbach_setup_tiles(jg)
+    END IF
+    CALL jsbach_init(jg)
+    CALL set_tc_current_date('1979-01-01T00:04:00Z')
+    !===================== END MANUALLY ADDED FOR FTG =====================!
     
     CALL ftg_destroy_savepoint()
     
