diff --git a/src/io/shared/mo_read_netcdf_distributed.f90 b/src/io/shared/mo_read_netcdf_distributed.f90
index bd8e4e5c8..bea096693 100644
--- a/src/io/shared/mo_read_netcdf_distributed.f90
+++ b/src/io/shared/mo_read_netcdf_distributed.f90
@@ -36,7 +36,11 @@ MODULE mo_read_netcdf_distributed
 
   IMPLICIT NONE
 
-  PRIVATE
+  !================= START FORCE FORTRAN TEST GENERATOR (FTG) =======================!
+
+  !PRIVATE
+
+  !================= STOP FORCE FORTRAN TEST GENERATOR (FTG) =========================!
 
   PUBLIC :: distrib_read
   PUBLIC :: distrib_nf_open
@@ -51,6 +55,13 @@ MODULE mo_read_netcdf_distributed
   PUBLIC :: distrib_inq_var_dims
   PUBLIC :: idx_lvl_blk, idx_blk_time
   PUBLIC :: nf
+  !================= START FORCE FORTRAN TEST GENERATOR (FTG) =======================!
+
+  PUBLIC :: get_empty_basic_distrib_read_data
+  PUBLIC :: t_basic_distrib_read_data
+  PUBLIC :: basic_data
+
+  !================= STOP FORCE FORTRAN TEST GENERATOR (FTG) =========================!
 
   INCLUDE 'netcdf.inc'
 
diff --git a/src/parallel_infrastructure/mo_communication.f90 b/src/parallel_infrastructure/mo_communication.f90
index 39c02cca4..14421086d 100644
--- a/src/parallel_infrastructure/mo_communication.f90
+++ b/src/parallel_infrastructure/mo_communication.f90
@@ -92,7 +92,11 @@ PUBLIC :: ASSIGNMENT(=)
 !
 TYPE t_comm_pattern
 
-  PRIVATE
+  !================= START FORCE FORTRAN TEST GENERATOR (FTG) =======================!
+
+  !PRIVATE
+
+  !================= STOP FORCE FORTRAN TEST GENERATOR (FTG) =========================!
 
    ! Number of points we receive in communication,
    ! this is the same as recv_limits
