diff --git a/src/configure_model/mo_mpi_phy_config.f90 b/src/configure_model/mo_mpi_phy_config.f90
index 11e5510ff..03f7d9f55 100644
--- a/src/configure_model/mo_mpi_phy_config.f90
+++ b/src/configure_model/mo_mpi_phy_config.f90
@@ -188,6 +188,14 @@ MODULE mo_mpi_phy_config
   !!
   TYPE(timedelta) , POINTER :: dt_zero
 
+
+  ! ================= BEGIN FORTRAN TEST GENERATOR (FTG) =======================
+
+  PUBLIC :: t_mpi_phy_config
+
+  ! ================= END FORTRAN TEST GENERATOR (FTG) =========================
+
+
 CONTAINS
 
   !----
diff --git a/src/io/shared/mo_read_netcdf_distributed.f90 b/src/io/shared/mo_read_netcdf_distributed.f90
index bd8e4e5c8..c5cbba3e0 100644
--- a/src/io/shared/mo_read_netcdf_distributed.f90
+++ b/src/io/shared/mo_read_netcdf_distributed.f90
@@ -36,7 +36,7 @@ MODULE mo_read_netcdf_distributed
 
   IMPLICIT NONE
 
-  PRIVATE
+  !PRIVATE
 
   PUBLIC :: distrib_read
   PUBLIC :: distrib_nf_open
@@ -52,6 +52,10 @@ MODULE mo_read_netcdf_distributed
   PUBLIC :: idx_lvl_blk, idx_blk_time
   PUBLIC :: nf
 
+  PUBLIC :: get_empty_basic_distrib_read_data
+  PUBLIC :: t_basic_distrib_read_data
+  PUBLIC :: basic_data
+
   INCLUDE 'netcdf.inc'
 
   INTERFACE distrib_read
diff --git a/src/parallel_infrastructure/mo_communication.f90 b/src/parallel_infrastructure/mo_communication.f90
index 39c02cca4..2a461695f 100644
--- a/src/parallel_infrastructure/mo_communication.f90
+++ b/src/parallel_infrastructure/mo_communication.f90
@@ -92,7 +92,7 @@ MODULE mo_communication
 !
 TYPE t_comm_pattern
 
-  PRIVATE
+  !PRIVATE
 
    ! Number of points we receive in communication,
    ! this is the same as recv_limits
